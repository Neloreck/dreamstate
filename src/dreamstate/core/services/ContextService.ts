import { queryData } from "@/dreamstate/core/queries/queryData";
import { emitSignal } from "@/dreamstate/core/signals/emitSignal";
import {
  IBaseSignal,
  IOptionalQueryRequest,
  TAnyObject,
  TDreamstateService,
  TQueryType,
  TSignalType
} from "@/dreamstate/types";

export abstract class ContextService<S extends TAnyObject = TAnyObject> {

  /**
   * Should core destroy store instance after observers removal or preserve it for application lifespan.
   * Singleton objects will never be destroyed once created.
   * Non-singleton objects are destroyed if all observers are removed.
   */
  protected static IS_SINGLE: boolean = false;

  public constructor(initialState?: S) {
  }

  /**
   * Lifecycle.
   * First provider was injected into DOM.
   */
  protected onProvisionStarted(): void {}

  /**
   * Lifecycle.
   * Last provider was removed from DOM.
   */
  protected onProvisionEnded(): void {}

  /**
   * Emit signal for other managers and subscribers.
   */
  protected emitSignal<T extends TSignalType = TSignalType, D = undefined>(
    baseSignal: IBaseSignal<T, D>
  ): Promise<void> {
    return emitSignal(baseSignal, this.constructor as TDreamstateService<S>);
  }

  /**
   * Send context query to retrieve data from @OnQuery method with required types.
   */
  protected queryData<
    R extends any,
    D extends any,
    T extends TQueryType,
    Q extends IOptionalQueryRequest<D, T> | Array<IOptionalQueryRequest>
    >(
    queryRequest: Q
  ) {
    return queryData<R, D, T, Q>(queryRequest);
  }

  /**
   * Send sync context query to retrieve data from @OnQuery method with required types.
   */
  protected queryDataSync<
    R extends any,
    D extends any,
    T extends TQueryType,
    Q extends IOptionalQueryRequest<D, T>
    >(
    queryRequest: Q
  ) {
    return queryData<R, D, T, Q>(queryRequest);
  }

}
