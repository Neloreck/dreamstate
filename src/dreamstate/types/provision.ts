import type { ReactNode } from "react";

/**
 * Properties for the Dreamstate context manager provider.
 * Defines the props used for initializing and configuring the provider.
 */
export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}
