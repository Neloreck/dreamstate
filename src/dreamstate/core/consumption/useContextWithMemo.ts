import { MutableRefObject, useCallback, useEffect, useRef, useState } from "react";

import { CONTEXT_STATES_REGISTRY, EMPTY_ARR } from "@/dreamstate/core/internals";
import { subscribeToManager } from "@/dreamstate/core/registry/subscribeToManager";
import { unsubscribeFromManager } from "@/dreamstate/core/registry/unSubscribeFromManager";
import { IContextManagerConstructor, TUpdateSubscriber } from "@/dreamstate/types";

/**
 * Use manager hook with subscribed updates.
 * Same like common useContext hook, but has memo checks.
 */
export function useContextWithMemo<T extends object, D extends IContextManagerConstructor<T>>(
  Manager: D,
  depsSelector: (context: T) => Array<any>
): T {
  const [ state, setState ] = useState(function() {
    return CONTEXT_STATES_REGISTRY.get(Manager) as T;
  });
  const observed: MutableRefObject<Array<any>> = useRef(depsSelector(state));

  const updateMemoState: TUpdateSubscriber<T> = useCallback(function(nextContext: T): void {
    // Calculate changes like react core does and fire change only after update.
    const nextObserved = depsSelector(nextContext);

    for (let it = 0; it < nextObserved.length; it ++) {
      if (observed.current[it] !== nextObserved[it]) {
        observed.current = nextObserved;
        setState(nextContext);

        return;
      }
    }
  }, EMPTY_ARR);

  useEffect(function() {
    subscribeToManager(Manager, updateMemoState);

    return function() {
      unsubscribeFromManager(Manager, updateMemoState);
    };
  });

  return state;
}
