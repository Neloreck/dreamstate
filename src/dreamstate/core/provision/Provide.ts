import { default as hoistNonReactStatics } from "hoist-non-react-statics";
import { ComponentType } from "react";

import { createManagersObserver } from "@/dreamstate/core/observing/createManagersObserver";
import { createClassWrapDecorator } from "@/dreamstate/polyfills/createClassWrapDecorator";
import { TDreamstateService } from "@/dreamstate/types";

/**
 * Decorator factory.
 * Provide context from context managers.
 * Observes changes and uses default react Providers for data flow.
 */
export function Provide<T extends ComponentType>(
  sources: Array<TDreamstateService<any>>
): ClassDecorator {
  return createClassWrapDecorator(function(targetClass: T): T {
    return hoistNonReactStatics(createManagersObserver(targetClass, sources), targetClass);
  });
}
