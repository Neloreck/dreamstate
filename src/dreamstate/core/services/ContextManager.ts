import { debug } from "@/macroses/debug.macro";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { notifyObservers } from "@/dreamstate/core/observing/notifyObservers";
import { shouldObserversUpdate } from "@/dreamstate/core/observing/shouldObserversUpdate";
import { getReactContext } from "@/dreamstate/core/registry/getReactContext";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import { TConstructorKey, TDreamstateService, TPartialTransformer } from "@/dreamstate/types";

/**
 * Abstract class.
 * Class based context manager for react.
 * Current Issue: Static items inside of each class instance.
 */
export abstract class ContextManager<T extends object> extends ContextService {

  /**
   * Lazy initialization, even for static resolving before anything from ContextManager is used.
   * Allows to get related React.Context for manual renders.
   */
  public static get REACT_CONTEXT() {
    if (this === ContextManager) {
      throw new Error("Cannot reference to ContextManager statics directly. Only inherited classes allowed.");
    }

    return getReactContext(this as TConstructorKey);
  }

  /**
   * Abstract store/actions bundle.
   */
  public abstract context: T;

  /**
   * Forces update and render of subscribed components.
   */
  public forceUpdate(): void {
    debug.info("Forcing context manager update:", this.constructor.name);
    // Force updates and common lifecycle with same params.
    this.beforeUpdate(this.context);
    this.context = Object.assign({}, this.context);
    notifyObservers(this);
    this.afterUpdate(this.context);
  }

  /**
   * Update current context from partially supplied state.
   * Calls lifecycle methods.
   */
  public setContext(next: Partial<T> | TPartialTransformer<T>): void {
    const previousContext: T = this.context;
    const nextContext: T = Object.assign(
      {},
      previousContext,
      typeof next === "function" ? next(previousContext) : next
    );

    /**
     * Compare current context with saved for observing one.
     */
    if (
      shouldObserversUpdate(
        CONTEXT_STATES_REGISTRY.get(this.constructor as TDreamstateService)!,
        nextContext
      )
    ) {
      debug.info("Updating context manager:", this.constructor.name);

      this.beforeUpdate(nextContext);
      this.context = nextContext;
      notifyObservers(this);
      this.afterUpdate(previousContext);
    } else {
      debug.info("Context manager update canceled, state is same:", this.constructor.name);
    }
  }

  /**
   * Lifecycle.
   * Before update lifecycle event.
   * Also shared for 'getSetter' methods.
   */
  protected beforeUpdate(nextContext: T): void {}

  /**
   * Lifecycle.
   * After update lifecycle event.
   * Also shared for 'getSetter' methods.
   */
  protected afterUpdate(previousContext: T): void {}

}
