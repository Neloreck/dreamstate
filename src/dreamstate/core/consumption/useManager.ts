import { useContext } from "react";

import { useContextWithMemo } from "@/dreamstate/core/consumption/useContextWithMemo";
import { IContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Use manager hook, higher order wrapper for useContext.
 */
export function useManager<
  T extends TAnyObject,
  D extends IContextManagerConstructor<T, any>
>(
  Manager: D,
  depsSelector?: (context: D["prototype"]["context"]) => Array<any>
): D["prototype"]["context"] {
  if (depsSelector) {
    return useContextWithMemo(Manager, depsSelector);
  } else {
    return useContext(Manager.REACT_CONTEXT);
  }
}
