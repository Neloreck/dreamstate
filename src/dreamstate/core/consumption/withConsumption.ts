import { Consume } from "@/dreamstate/core/consumption/Consume";
import { IConsume } from "@/dreamstate/types";

/**
 * HOC alias for @Consume.
 */
export const withConsumption: IConsume = Consume as IConsume;
