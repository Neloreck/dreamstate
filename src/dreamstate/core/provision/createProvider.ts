import { FunctionComponent, ReactNode, useCallback, useState } from "react";

import { EMPTY_ARR } from "@/dreamstate/core/internals";
import { useHotObservers } from "@/dreamstate/core/observing/useHotObservers";
import { useStaticObservers } from "@/dreamstate/core/observing/useStaticObservers";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import {
  TAnyContextManagerConstructor,
  TAnyContextServiceConstructor,
  TAnyObject
} from "@/dreamstate/types";

export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider<
  T extends TAnyObject = TAnyObject
>(
  sources: Array<TAnyContextServiceConstructor>
): FunctionComponent<IProviderProps<T>> {
  if (!Array.isArray(sources)) {
    throw new TypeError(
      "Wrong providers parameter supplied. Only array of context services is acceptable."
    );
  }

  for (let it = 0; it < sources.length; it ++) {
    if (!sources[it] || !(sources[it].prototype instanceof ContextService)) {
      throw new TypeError("Only classes extending ContextService can be supplied for provision.");
    }
  }

  /**
   * Check only managers with required provision.
   * Do not include services for subTree rendering but add registering logic for services.
   */
  const managers: Array<TAnyContextManagerConstructor> = sources.filter(function(
    Service: TAnyContextServiceConstructor
  ) {
    return Service.prototype instanceof ContextManager;
  }) as Array<TAnyContextManagerConstructor>;

  /**
   * Create observer component that will handle observing.
   */
  function Observer(props: IProviderProps<T>): ReactNode {
    const [ , forceRender ] = useState({});
    const updateProviders = useCallback(function() {
      forceRender({});
    }, EMPTY_ARR);

    if (props.partialHotReplacement) {
      useHotObservers(sources, props.initialState, updateProviders);
    } else {
      useStaticObservers(sources, props.initialState, updateProviders);
    }

    return provideSubTreeRecursive(props.children, managers, 0);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextServiceConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
