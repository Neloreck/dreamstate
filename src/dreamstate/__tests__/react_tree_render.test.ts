import { mount } from "enzyme";
import { createElement, PureComponent, ReactElement, ReactNode } from "react";

import { Consume } from "@/dreamstate/core/consumption/Consume";
import { useManager } from "@/dreamstate/core/consumption/useManager";
import { createProvider } from "@/dreamstate/core/provision/createProvider";
import { Provide } from "@/dreamstate/core/provision/Provide";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";

describe("React tree for provided and consumed components", () => {
  interface IExampleContext {
    exampleNumber: number;
    exampleString: string;
    exampleActions: {
      setExampleNumber(value: number): void;
      setExampleString(value: string): void;
    };
  }

  class ExampleContextManager extends ContextManager<IExampleContext> {

    public readonly context: IExampleContext = {
      exampleActions: {
        setExampleNumber: this.setExampleNumber.bind(this),
        setExampleString: this.setExampleString.bind(this)
      },
      exampleNumber: 0,
      exampleString: "initial"
    };

    public setExampleNumber(value: number): void {
      this.setContext({ exampleNumber: value });
    }

    public setExampleString(value: string): void {
      this.setContext({ exampleString: value });
    }

  }

  class ProvidedService extends ContextService {
  }

  interface IPlaceholderContext {
    placeholder: true;
  }

  class PlaceholderContextManager extends ContextManager<IPlaceholderContext> {

    public readonly context: IPlaceholderContext = {
      placeholder: true
    };

  }

  const Provider = createProvider([ ExampleContextManager, PlaceholderContextManager, ProvidedService ]);

  const ExampleContextFunctionalProvider = Provider;

  @Provide([ ExampleContextManager, PlaceholderContextManager, ProvidedService ])
  class ExampleContextDecoratedProvider extends PureComponent {

    public render(): ReactNode {
      return this.props.children;
    }

  }

  function ExampleContextFunctionalConsumer(): ReactElement {
    const value: IExampleContext = useManager(ExampleContextManager);

    return createElement("span", {}, JSON.stringify(value));
  }

  @Consume([ { from: ExampleContextManager, as: "exampleContext" } ])
  class ExampleContextClassConsumer extends PureComponent<{
    exampleContext: IExampleContext;
  }> {

    public render(): ReactNode {
      return createElement("span", {}, JSON.stringify(this.props.exampleContext));
    }

  }

  it("Should correctly provide manager context from functional provider", () => {
    const functionalFunctionalTree = mount(
      createElement(ExampleContextFunctionalProvider, {} as any, createElement(ExampleContextFunctionalConsumer))
    );

    expect(functionalFunctionalTree).toMatchSnapshot("Provision: Functional - Functional.");

    const functionalDecoratedTree = mount(
      createElement(ExampleContextFunctionalProvider, {} as any, createElement(ExampleContextClassConsumer))
    );

    expect(functionalDecoratedTree).toMatchSnapshot("Provision: Functional - Decorated.");
  });

  it("Should correctly provide manager context from decorated HoC provider", () => {
    const decoratedFunctionalTree = mount(
      createElement(ExampleContextDecoratedProvider, {} as any, createElement(ExampleContextFunctionalConsumer))
    );

    expect(decoratedFunctionalTree).toMatchSnapshot("Provision: Decorated - Functional.");

    const decoratedDecoratedTree = mount(
      createElement(ExampleContextDecoratedProvider, {} as any, createElement(ExampleContextClassConsumer))
    );

    expect(decoratedDecoratedTree).toMatchSnapshot("Provision: Decorated - Decorated.");
  });
});
