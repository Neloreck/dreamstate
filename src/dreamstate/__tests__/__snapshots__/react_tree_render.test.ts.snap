// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React tree for provided and consumed components. Should correctly provide manager context from decorated HoC provider.: Provision: Decorated - Decorated. 1`] = `
<Memo(DS.Observer)>
  <ExampleContextDecoratedProvider>
    <DS.Consumer>
      <ExampleContextClassConsumer
        exampleContext={
          Object {
            "exampleActions": Object {
              "setExampleNumber": [Function],
              "setExampleString": [Function],
            },
            "exampleNumber": 0,
            "exampleString": "initial",
          }
        }
      >
        <span>
          {"exampleActions":{},"exampleNumber":0,"exampleString":"initial"}
        </span>
      </ExampleContextClassConsumer>
    </DS.Consumer>
  </ExampleContextDecoratedProvider>
</Memo(DS.Observer)>
`;

exports[`React tree for provided and consumed components. Should correctly provide manager context from decorated HoC provider.: Provision: Decorated - Functional. 1`] = `
<Memo(DS.Observer)>
  <ExampleContextDecoratedProvider>
    <ExampleContextFunctionalConsumer>
      <span>
        {"exampleActions":{},"exampleNumber":0,"exampleString":"initial"}
      </span>
    </ExampleContextFunctionalConsumer>
  </ExampleContextDecoratedProvider>
</Memo(DS.Observer)>
`;

exports[`React tree for provided and consumed components. Should correctly provide manager context from functional provider.: Provision: Functional - Decorated. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer>
    <ExampleContextClassConsumer
      exampleContext={
        Object {
          "exampleActions": Object {
            "setExampleNumber": [Function],
            "setExampleString": [Function],
          },
          "exampleNumber": 0,
          "exampleString": "initial",
        }
      }
    >
      <span>
        {"exampleActions":{},"exampleNumber":0,"exampleString":"initial"}
      </span>
    </ExampleContextClassConsumer>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`React tree for provided and consumed components. Should correctly provide manager context from functional provider.: Provision: Functional - Functional. 1`] = `
<Memo(DS.Observer)>
  <ExampleContextFunctionalConsumer>
    <span>
      {"exampleActions":{},"exampleNumber":0,"exampleString":"initial"}
    </span>
  </ExampleContextFunctionalConsumer>
</Memo(DS.Observer)>
`;
